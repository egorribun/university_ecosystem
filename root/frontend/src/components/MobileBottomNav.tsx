import { NavLink, useLocation } from "react-router-dom";
import { useLayoutEffect, useMemo } from "react";
import DashboardIcon from "@mui/icons-material/Dashboard";
import ArticleIcon from "@mui/icons-material/Article";
import EventNoteIcon from "@mui/icons-material/EventNote";
import TodayIcon from "@mui/icons-material/Today";
import PersonIcon from "@mui/icons-material/Person";

function getScrollRoot(): HTMLElement {
  const cands: (Element | null | Document | HTMLElement)[] = [
    document.querySelector("[data-scroll-root]"),
    document.querySelector("main[role='main']"),
    document.querySelector("main"),
    document.getElementById("scroll-root"),
    document.querySelector("#root"),
    (document as any).scrollingElement,
    document.documentElement,
    document.body
  ];
  for (const el of cands) {
    if (!el) continue;
    const e = el as HTMLElement;
    const oy = getComputedStyle(e).overflowY;
    const scrollable = (oy === "auto" || oy === "scroll") && e.scrollHeight > e.clientHeight;
    if (scrollable) return e;
  }
  return (document.scrollingElement || document.documentElement) as HTMLElement;
}

function smoothToTop(target: HTMLElement) {
  try {
    (target as any).scrollTo({ top: 0, behavior: "smooth" });
  } catch {
    const start = target.scrollTop;
    const duration = 420;
    let t0 = 0;
    const step = (ts: number) => {
      if (!t0) t0 = ts;
      const p = Math.min(1, (ts - t0) / duration);
      const eased = 1 - Math.pow(1 - p, 3);
      target.scrollTop = Math.round(start * (1 - eased));
      if (p < 1) requestAnimationFrame(step);
    };
    requestAnimationFrame(step);
  }
}

function markIfFromBottom() {
  const el = getScrollRoot();
  const threshold = 24;
  const nearBottom = el.scrollTop + el.clientHeight >= el.scrollHeight - threshold;
  if (nearBottom) sessionStorage.setItem("__scrollTopNext", "1");
}

function samePath(a: string, b: string) {
  const na = a.replace(/\/+$/, "") || "/";
  const nb = b.replace(/\/+$/, "") || "/";
  return na === nb;
}

export default function MobileBottomNav() {
  const { pathname } = useLocation();

  useLayoutEffect(() => {
    if (sessionStorage.getItem("__scrollTopNext") === "1") {
      sessionStorage.removeItem("__scrollTopNext");
      const el = getScrollRoot();
      requestAnimationFrame(() => requestAnimationFrame(() => smoothToTop(el)));
    }
  }, [pathname]);

  const items = useMemo(
    () => [
      { to: "/dashboard", label: "Главная", icon: <DashboardIcon /> },
      { to: "/news", label: "Новости", icon: <ArticleIcon /> },
      { to: "/events", label: "События", icon: <EventNoteIcon /> },
      { to: "/schedule", label: "Расписание", icon: <TodayIcon /> },
      { to: "/profile", label: "Профиль", icon: <PersonIcon /> }
    ],
    []
  );

  const hideOn = ["/login", "/register", "/forgot-password", "/reset-password"];
  const hidden = hideOn.some((p) => pathname.startsWith(p));
  if (hidden) return null;

  return (
    <>
      <nav className="bottom-nav glass" role="navigation" aria-label="Основная навигация">
        {items.map((it) => (
          <NavLink
            key={it.to}
            to={it.to}
            onPointerDown={markIfFromBottom}
            onClick={(e) => {
              if (samePath(pathname, it.to)) {
                e.preventDefault();
                const el = getScrollRoot();
                requestAnimationFrame(() => smoothToTop(el));
              }
            }}
            onKeyDown={(e) => {
              if ((e.key === "Enter" || e.key === " ") && samePath(pathname, it.to)) {
                e.preventDefault();
                const el = getScrollRoot();
                requestAnimationFrame(() => smoothToTop(el));
              }
            }}
            className={({ isActive }) => "bottom-nav__item" + (isActive ? " active" : "")}
            aria-label={it.label}
          >
            <span className="bottom-nav__icon">{it.icon}</span>
            <span className="bottom-nav__label">{it.label}</span>
          </NavLink>
        ))}
      </nav>
      <div className="bottom-nav-spacer" aria-hidden="true" />
    </>
  );
}