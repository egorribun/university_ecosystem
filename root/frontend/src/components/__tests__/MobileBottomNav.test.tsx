import { MemoryRouter } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import MobileBottomNav from "../MobileBottomNav";

describe("MobileBottomNav", () => {
  it("does not render on auth pages", () => {
    render(
      <MemoryRouter initialEntries={["/login"]}>
        <MobileBottomNav />
      </MemoryRouter>
    );

    expect(screen.queryByRole("navigation", { name: "Основная навигация" })).toBeNull();
  });

  it("renders links for main sections", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/dashboard"]}>
        <MobileBottomNav />
      </MemoryRouter>
    );

    const nav = screen.getByRole("navigation", { name: "Основная навигация" });
    expect(nav).toBeInTheDocument();
    expect(nav.querySelectorAll("a")).toHaveLength(5);
    expect(nav).toMatchSnapshot();
    expect(container).toMatchSnapshot();
  });
});
