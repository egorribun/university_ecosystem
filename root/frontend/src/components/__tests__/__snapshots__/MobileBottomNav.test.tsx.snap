// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MobileBottomNav > renders links for main sections 1`] = `
<nav
  aria-label="Основная навигация"
  class="bottom-nav glass"
  role="navigation"
>
  <a
    aria-current="page"
    aria-label="Главная"
    class="bottom-nav__item active"
    href="/dashboard"
  >
    <span
      class="bottom-nav__icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="DashboardIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 13h8V3H3zm0 8h8v-6H3zm10 0h8V11h-8zm0-18v6h8V3z"
        />
      </svg>
    </span>
    <span
      class="bottom-nav__label"
    >
      Главная
    </span>
  </a>
  <a
    aria-label="Новости"
    class="bottom-nav__item"
    href="/news"
  >
    <span
      class="bottom-nav__icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="ArticleIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m-5 14H7v-2h7zm3-4H7v-2h10zm0-4H7V7h10z"
        />
      </svg>
    </span>
    <span
      class="bottom-nav__label"
    >
      Новости
    </span>
  </a>
  <a
    aria-label="События"
    class="bottom-nav__item"
    href="/events"
  >
    <span
      class="bottom-nav__icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="EventNoteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 10H7v2h10zm2-7h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V8h14zm-5-5H7v2h7z"
        />
      </svg>
    </span>
    <span
      class="bottom-nav__label"
    >
      События
    </span>
  </a>
  <a
    aria-label="Расписание"
    class="bottom-nav__item"
    href="/schedule"
  >
    <span
      class="bottom-nav__icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="TodayIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V8h14zM7 10h5v5H7z"
        />
      </svg>
    </span>
    <span
      class="bottom-nav__label"
    >
      Расписание
    </span>
  </a>
  <a
    aria-label="Профиль"
    class="bottom-nav__item"
    href="/profile"
  >
    <span
      class="bottom-nav__icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="PersonIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
        />
      </svg>
    </span>
    <span
      class="bottom-nav__label"
    >
      Профиль
    </span>
  </a>
</nav>
`;

exports[`MobileBottomNav > renders links for main sections 2`] = `
<div>
  <nav
    aria-label="Основная навигация"
    class="bottom-nav glass"
    role="navigation"
  >
    <a
      aria-current="page"
      aria-label="Главная"
      class="bottom-nav__item active"
      href="/dashboard"
    >
      <span
        class="bottom-nav__icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="DashboardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 13h8V3H3zm0 8h8v-6H3zm10 0h8V11h-8zm0-18v6h8V3z"
          />
        </svg>
      </span>
      <span
        class="bottom-nav__label"
      >
        Главная
      </span>
    </a>
    <a
      aria-label="Новости"
      class="bottom-nav__item"
      href="/news"
    >
      <span
        class="bottom-nav__icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArticleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m-5 14H7v-2h7zm3-4H7v-2h10zm0-4H7V7h10z"
          />
        </svg>
      </span>
      <span
        class="bottom-nav__label"
      >
        Новости
      </span>
    </a>
    <a
      aria-label="События"
      class="bottom-nav__item"
      href="/events"
    >
      <span
        class="bottom-nav__icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="EventNoteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 10H7v2h10zm2-7h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V8h14zm-5-5H7v2h7z"
          />
        </svg>
      </span>
      <span
        class="bottom-nav__label"
      >
        События
      </span>
    </a>
    <a
      aria-label="Расписание"
      class="bottom-nav__item"
      href="/schedule"
    >
      <span
        class="bottom-nav__icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="TodayIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V8h14zM7 10h5v5H7z"
          />
        </svg>
      </span>
      <span
        class="bottom-nav__label"
      >
        Расписание
      </span>
    </a>
    <a
      aria-label="Профиль"
      class="bottom-nav__item"
      href="/profile"
    >
      <span
        class="bottom-nav__icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
          />
        </svg>
      </span>
      <span
        class="bottom-nav__label"
      >
        Профиль
      </span>
    </a>
  </nav>
  <div
    aria-hidden="true"
    class="bottom-nav-spacer"
  />
</div>
`;
